File: ios/chillnote/Services/AppsFlyerService.swift
import AppsFlyerLib
import Foundation
import OSLog
import UIKit

struct AppsFlyerInstallAttribution {
    let mediaSource: String?
    let campaign: String?
    let adGroup: String?
    let ad: String?

    init(conversionData: [AnyHashable: Any]) {
        func string(_ key: String) -> String? {
            guard let value = conversionData[key] else { return nil }
            let normalized = String(describing: value)
                .trimmingCharacters(in: .whitespacesAndNewlines)
            return normalized.isEmpty ? nil : normalized
        }

        mediaSource = string("media_source")
        campaign = string("campaign")
        adGroup = string("af_adset")
        ad = string("af_ad")
    }
}

final class AppsFlyerService: NSObject, AppsFlyerLibDelegate {
    static let shared = AppsFlyerService()

    private static let appleAppID = "6758427839"
    private static let logger = Logger(subsystem: "com.chillnote.app", category: "appsflyer")
    private var isConfigured = false

    private override init() {}

    @MainActor
    func configure(launchOptions: [UIApplication.LaunchOptionsKey: Any]?) {
        guard !isConfigured else { return }
        guard let devKey = AppConfig.appsFlyerDevKey else {
            Self.logger.notice("AppsFlyer is disabled because APPSFLYER_DEV_KEY is missing")
            return
        }

        let appsFlyer = AppsFlyerLib.shared()
        appsFlyer.initialize(devKey: devKey, appId: Self.appleAppID)
        appsFlyer.delegate = self
        appsFlyer.customerUserID = AuthService.shared.currentUserId.map(RevenueCatIdentity.canonicalUserID)
        appsFlyer.handleLaunchOptions(launchOptions)
#if DEBUG
        appsFlyer.isDebug = true
#endif
        appsFlyer.registerSessionReadyListener { [weak self] in
            AppsFlyerLib.shared().start()
            self?.syncDeviceIdentifiersToRevenueCat()
        }
        isConfigured = true
    }

    @MainActor
    func identify(userID: String?) {
        guard isConfigured else { return }
        AppsFlyerLib.shared().customerUserID = userID.map(RevenueCatIdentity.canonicalUserID)
        syncDeviceIdentifiersToRevenueCat()
    }

    func onConversionDataSuccess(_ conversionData: [AnyHashable: Any]) {
        let status = String(describing: conversionData["af_status"] ?? "")
        let attribution = status.caseInsensitiveCompare("Non-organic") == .orderedSame
            ? AppsFlyerInstallAttribution(conversionData: conversionData)
            : nil

        Task { @MainActor in
            RevenueCatService.shared.setAppsFlyerAttribution(
                appsFlyerID: AppsFlyerLib.shared().getAppsFlyerUID(),
                attribution: attribution
            )
        }
    }

    func onConversionDataFail(_ error: Error) {
        Self.logger.warning(
            "AppsFlyer conversion data failed: \(error.localizedDescription, privacy: .public)"
        )
        syncDeviceIdentifiersToRevenueCat()
    }

    private func syncDeviceIdentifiersToRevenueCat() {
        let appsFlyerID = AppsFlyerLib.shared().getAppsFlyerUID()
        Task { @MainActor in
            RevenueCatService.shared.setAppsFlyerAttribution(
                appsFlyerID: appsFlyerID,
                attribution: nil
            )
        }
    }
}

File: android/app/src/main/java/com/sponteoai/chillscript/analytics/AppsFlyerService.kt
package com.sponteoai.chillscript.analytics

import android.content.Context
import android.util.Log
import com.appsflyer.AppsFlyerLib
import com.appsflyer.share.AppsFlyerConversionListener
import com.revenuecat.purchases.Purchases
import com.sponteoai.chillscript.BuildConfig

data class AppsFlyerInstallAttribution(
    val mediaSource: String?,
    val campaign: String?,
    val adGroup: String?,
    val ad: String?,
) {
    companion object {
        fun from(conversionData: Map<String, Any>): AppsFlyerInstallAttribution =
            AppsFlyerInstallAttribution(
                mediaSource = conversionData.stringValue("media_source"),
                campaign = conversionData.stringValue("campaign"),
                adGroup = conversionData.stringValue("af_adset"),
                ad = conversionData.stringValue("af_ad"),
            )

        private fun Map<String, Any>.stringValue(key: String): String? =
            get(key)?.toString()?.trim()?.takeIf(String::isNotEmpty)
    }
}

object AppsFlyerService {
    private const val TAG = "AppsFlyerService"
    private var isConfigured = false

    fun configure(context: Context) {
        if (isConfigured) return
        val devKey = BuildConfig.APPSFLYER_DEV_KEY.trim()
        if (devKey.isEmpty() || devKey.contains("REPLACE", ignoreCase = true)) {
            Log.i(TAG, "AppsFlyer is disabled because its developer key is missing")
            return
        }

        val appContext = context.applicationContext
        val listener = object : AppsFlyerConversionListener {
            override fun onConversionDataSuccess(conversionData: Map<String, Any>) {
                val data = conversionData
                val attribution = if (
                    data["af_status"]?.toString().equals("Non-organic", ignoreCase = true)
                ) {
                    AppsFlyerInstallAttribution.from(data)
                } else {
                    null
                }
                syncRevenueCat(appContext, attribution)
            }

            override fun onConversionDataFail(errorMessage: String) {
                Log.w(TAG, "AppsFlyer conversion data failed: $errorMessage")
                syncRevenueCat(appContext, null)
            }
        }

        val appsFlyer = AppsFlyerLib.getInstance()
        appsFlyer.init(devKey, listener, appContext)
        if (BuildConfig.DEBUG) appsFlyer.setDebugLog(true)
        appsFlyer.registerSessionReadyListener {
            appsFlyer.start()
            syncRevenueCat(appContext, null)
        }
        isConfigured = true
    }

    fun identify(context: Context, userID: String?) {
        if (!isConfigured || userID.isNullOrBlank()) return
        AppsFlyerLib.getInstance().setCustomerUserId(userID.lowercase())
        syncRevenueCat(context.applicationContext, null)
    }

    private fun syncRevenueCat(
        context: Context,
        attribution: AppsFlyerInstallAttribution?,
    ) {
        if (!Purchases.isConfigured) return
        val purchases = Purchases.sharedInstance
        purchases.collectDeviceIdentifiers()
        AppsFlyerLib.getInstance().getAppsFlyerUID(context)
            ?.takeIf(String::isNotBlank)
            ?.let(purchases::setAppsflyerID)
        attribution?.mediaSource?.let(purchases::setMediaSource)
        attribution?.campaign?.let(purchases::setCampaign)
        attribution?.adGroup?.let(purchases::setAdGroup)
        attribution?.ad?.let(purchases::setAd)
    }
}
